qadence/constructors/hamiltonians.py

```diff
@@ -119,9 +119,16 @@ def hamiltonian_factory(
 
     # Get interaction function
     if interaction is not None:
-        int_fn = INTERACTION_DICT.get(interaction, None)
-        if int_fn is None:
-            raise KeyError(f"Interaction {interaction} not supported.")
+        if callable(interaction):
+            int_fn = interaction
+            try:
+                interaction(0, 1)
```

Comment:
I am no sure it makes for a solid test because it won't pick up if the function signature is type annotated with floats for instance. I would rather use `inspect` or something like [this](https://stackoverflow.com/questions/73206836/how-to-verify-the-signature-of-python-method-when-using-default-parameters).

Reply:
Well, my point being that if you want to _at least_ check that the signature is two integers then it makes sense to use `inspect`. Otherwise, there is no other way to ensure validity of the function imho.

Reply:
I guess Roland has  valid concern. However, we are writing Python code. And that would not be Pythonic.

Viewing Qadence as a DSL, it would in any case raise a "compile-time" error, since it would fail immediately. 

I'm not sure if it's a good idea to plan to introduce that level of type-safety in Python, since it goes against the way Python is intended to be used... Swimming upstream...

Reply:
I understand that this is a PITA and very likely an overkill. My point is, the only meaningful check is signature types. Otherwise, I think we can ignore it.

Reply:
Hey @Roland-djee I did try to use `inspect.signature` but it would amount to parsing a string like `"i: int"` to check if the second part was "int" or something else. It seemed a bit of a weird usage to me and maybe not very meaningful since it doesn't control what the function returns.

Then I tried adding a specific control of the return to check it is either Pauli terms or Projectors, but ended up running into some failing tests for Hamiltonians built on Projectors. Since I don't want to delay this, I just dropped the support for Projectors for now and made it so only Pauli operators are allowed, which either way will be most of the usage for now, and IMO the control now is the most robust.

I opened an issue here for adding the Projectors later: https://github.com/pasqal-io/qadence/issues/300

Reply:
OK fair enough. My main concern was primarily from the fact that a user can actually plug in a function and break everything. Nevertheless, I also understand that: 1) it's the user's responsibility to do things sensibly, especially if informed 2) it's Python and we can't do much about safety in Python.